command/connect/envoy/exec_unix.go

```diff
@@ -43,126 +43,94 @@ func hasHotRestartOption(argSets ...[]string) bool {
-	// `fd`. This is accessible as a file itself (trippy!) under
-	// /dev/fd/$FDNUMBER.
-	magicPath := filepath.Join("/dev/fd", strconv.Itoa(int(fd)))
+	pipeFile := filepath.Join(dir, "bootstrap.json")
 
-	// We default to disabling hot restart because it makes it easier to run
-	// multiple envoys locally for testing without them trying to share memory and
-	// unix sockets and complain about being different IDs. But if user is
-	// actually configuring hot-restart explicitly with the --restart-epoch option
-	// then don't disable it!
-	disableHotRestart := !hasHotRestartOption(prefixArgs, suffixArgs)
-
-	// First argument needs to be the executable name.
-	envoyArgs := []string{binary}
-	envoyArgs = append(envoyArgs, prefixArgs...)
-	envoyArgs = append(envoyArgs, "--v2-config-only",
-		"--config-path",
-		magicPath,
-	)
-	if disableHotRestart {
-		envoyArgs = append(envoyArgs, "--disable-hot-restart")
-	}
-	envoyArgs = append(envoyArgs, suffixArgs...)
-
-	// Exec
-	if err = unix.Exec(binary, envoyArgs, os.Environ()); err != nil {
-		return errors.New("Failed to exec envoy: " + err.Error())
-	}
-
-	return nil
-}
-
-func writeEphemeralEnvoyTempFile(b []byte) (uintptr, error) {
-	f, err := ioutil.TempFile("", "envoy-ephemeral-config")
+	err = syscall.Mkfifo(pipeFile, 0600)
 	if err != nil {
-		return 0, err
+		return dir, err
 	}
 
-	errFn := func(err error) (uintptr, error) {
-		_ = f.Close()
-		return 0, err
-	}
-
-	// TempFile already does this, but it's cheap to reinforce that we
-	// WANT the default behavior.
-	if err := f.Chmod(0600); err != nil {
-		return errFn(err)
+	// Create an anonymous pipe for STDIN to the child that will write to the
+	// named pipe.
+	pr, pw, err := os.Pipe()
+	if err != nil {
+		return dir, err
 	}
 
-	// Immediately unlink the file as we are going to just pass the
-	// file descriptor, not the path.
-	if err = os.Remove(f.Name()); err != nil {
-		return errFn(err)
-	}
-	if _, err = f.Write(b); err != nil {
-		return errFn(err)
+	// Write the config
+	n, err := pw.Write(bootstrapJSON)
+	if err != nil {
+		return dir, err
 	}
-	// Rewind the file descriptor so Envoy can read it.
-	if _, err = f.Seek(0, io.SeekStart); err != nil {
-		return errFn(err)
+	if n < len(bootstrapJSON) {
+		return dir, fmt.Errorf("failed writing boostrap to child STDIN: %s", err)
 	}
 
-	// Disable CLOEXEC so that this file descriptor is available
-	// to the exec'd Envoy.
-	if err := setCloseOnExec(f.Fd(), false); err != nil {
-		return errFn(err)
+	err = pw.Close()
+	if err != nil {
+		return dir, err
+	}
+
+	// Start a detached child process that we can securely send the bootstrap via
+	// STDIN and have it write to the pipe. Cat can do this.
+	var attr = os.ProcAttr{
+		Dir: ".",
+		Env: os.Environ(),
+		Files: []*os.File{
+			pr,
+			nil,
+			nil,
+		},
+	}
+	process, err := os.StartProcess("/bin/bash",
+		[]string{"bash", "-c", "cat > " + pipeFile + "; rm -rf " + pipeFile},
```

Comment:
You could also add an consul subcommand that did this operation rather than doing a blob of inline shell. That subcommand could be omitted from the general help listings as it isn't terribly useful on its own.

As is you make the assumption here that `/bin/bash` is valid and that `cat & rm` are both on the `PATH`. These are all reasonable assumptions, but new here.

Another bit of awfulness here, with inline shell (via `bash -c`) you now have to think a bit harder about whether `pipeFile` could have spaces in the full path. Normally no, but what if someone set `TMPDIR` to an absolute path with spaces in the name?  If you could manage this as a single subprocess invocation with the `pipeFile` value only manifesting in a single positional argument you wouldn't have to worry (like some mythical app named `cp_and_rm`).

Reply:
This pretty much amounts to code golf, but you can use `cp` instead of `cat` if that seems better:
```
$ echo foo | cp /dev/stdin my.out
$ cat my.out
foo
```

Reply:
Hmm yeah I considered a custom command. Requiring bash and cat seemed _ok_ to me but I don't really want to go handling issues for this on arcane minimal setups etc.